### PR TITLE
[Issue#51] Don't rename columns to PID if it already exists

### LINKED ISSUE #49: Columns in ACE output file renamed leading to error in processing
Problem: Column N and P of the ACE demographics file were renamed, which seems to be causing an error in processing

Error provided in the R script: _Error in if (mask$get_size() > 0) { : argument is of length zero_

Potential solution: Update aceR script to handle "Participant ID" for column N and "PID" for column Y of the demographics file. In previous versions they were set as "Nexus.Id" and "Participant.Id" respectively

### LINKED ISSUE #51: Error in loading demographics.csv
Problem: "I found the aceR toolbox cannot load the demographic.csv and will generate error if I remove the demographic.csv. Attached please kindly find the R code and the output." This may be related to #49 as the columns N and Y are with the new name convention
[ACE_RawData_WPI2-20240531T173409Z-001.zip](https://github.com/user-attachments/files/15516999/ACE_RawData_WPI2-20240531T173409Z-001.zip)
Attached is the file that they were trying to process

### PR DESCRIPTION
**Pull request by arjunpur**
This PR includes a fix for https://github.com/joaquinanguera/aceR/issues/51 & https://github.com/joaquinanguera/aceR/issues/49 (likely a duplicate of 51).

The demographics.csv now has a `PID` column that's populated by Nexus. This PR updates the `standardize_ace_column_names` function to only rename the `participant_id` and `user_id` columns IF the `PID` column doesn't already exist in the dataframe. 

The root of the issue was that we were creating a duplicate `PID` column which causes problems when we try to use `mutate`

### CODE REVIEW

**Comment by arjunpur** [comment]
@monicathieu when you get a chance, could you take a look at this?

File: R/load-ace-constant-columns.R [hunk 1]
```diff
@@ -206,9 +205,24 @@ standardize_ace_column_names <- function(df) {
                   new == "task_switch_state" ~ "taskswitch_state",
                   TRUE ~ new) # for cross compatibility b/w emailed and pulvinar)
   names(df) = new
+
+  # With newer data formats (nexus), "PID" may already be included alongside
+  # an existing "participant_id". In this case, do not map "participant_id"
+  # to "PID" as we'll be creating a duplicate
+  rename_conditionally(df, "participant_id", COL_PID)
+  rename_conditionally(df, "user_id", COL_PID)
```
Comment (monicathieu) [inline R/load-ace-constant-columns.R]: Quick question: Are these actually modifying `df` if they are not assigning the output of `rename_conditionally()` back to `df`? I realize there may yet be R behaviors I'm unaware of, so I figured I'd ask whether this definitely works or whether it ought to be:

```suggestion
  df <- rename_conditionally(df, "participant_id", COL_PID)
  df <- rename_conditionally(df, "user_id", COL_PID)
```

File: R/load-ace-constant-columns.R [hunk 2]
```diff
@@ -206,9 +205,24 @@ standardize_ace_column_names <- function(df) {
...
 }
 
+rename_conditionally <- function(df, old_name, new_name) {
+  if (old_name %in% names(df) && !(new_name %in% names(df))) {
+    df <- df %>% rename(!!new_name := !!sym(old_name))
+  }
+  df
+}
+
+
```
Comment (monicathieu) [inline R/load-ace-constant-columns.R]: This new function looks very handy! Would you mind moving this over to utils-standardize.R? I think it fits over there as `standardize_names()` is also defined over there.

**Review by monicathieu — COMMENTED** [review]
This looks like it should work overall! Thanks for chasing down the source of that issue. Please see two very small comments.

If it is indeed the case that the new Nexus `PID` column in the raw data contains the same information as `participant_id`, then this will work great as it doesn't matter which column is used. I approve merging these changes conditional on confirming that the new `PID` column is indeed the one to use.

**Comment by arjunpur** [comment]
> If it is indeed the case that the new Nexus PID column in the raw data contains the same information as participant_id, then this will work great as it doesn't matter which column is used. I approve merging these changes conditional on confirming that the new PID column is indeed the one to use.

I verified in the Nexus backend codebase that the `PID` column is indeed populated with the same information as `participantId`. It was added in 2022

image

File: R/load-ace-constant-columns.R (hunk 1 above)
Comment (arjunpur) [reply R/load-ace-constant-columns.R]: I think you're right actually. I verified that `df` is not getting modified despite being passed in. Will update this

File: R/load-ace-constant-columns.R (hunk 2 above)
Comment (arjunpur) [reply R/load-ace-constant-columns.R]: Will do!

**Comment by arjunpur** [comment]
Thanks for the comments @monicathieu - pushed up updates to address them

**Comment by monicathieu** [comment]
Looks good! I approve merging :)

PS: This is a nice bug fix that probably merits incrementing the version number and noting it the readme release notes! Since it's 2024 now we might even be able to jump up to 2024.0.0. I'd say this bug fix is technically worth only a micro version update, but since it goes along with adding Nexus support (which is certainly a major version update) I think the calendar versioning jump up also makes sense semantically.

**Comment by arjunpur** [comment]
Good call @monicathieu - I updated the DESCRIPTION to use `24.0.0` as the current version + updated the README.md. Will merge this now thanks!

**Pull request merged**